### DISCUSSION TITLE
Welcome to autokey Discussions!

### DISCUSSION THREAD

**Discussion by josephj11** [autokey/autokey#540] [General]
## 👋 Welcome!
  We’re using Discussions primarily as a place to track Beta testing and other project related work.

Whenever you have a question (as opposed to an issue) it's better to post it on our [list](https://groups.google.com/forum/#!forum/autokey-users) or on [Gitter](https://gitter.im/autokey/autokey). The list is more user oriented whereas more technical questions will get a better answer on Gitter.

Both of these venues have a number of active users, so you are likely to get more, better, and faster answers on them than here on GitHub where only our very busy developers and a few users see the posts. We also prefer to have the developers developing rather than taking time to answer questions and manage issues that the user community could handle for them.

**Comment by josephj11** [comment]
## Installing The Beta Version Via pip

Make a directory and change into it then

    pip3 install --user https://github.com/autokey/autokey/archive/v0.96.0-beta.1.zip

**Reply by BlueDrink9** [reply]
Update to `pip3 install --user https://github.com/autokey/autokey/archive/v0.96.0-beta.4.zip`

**Comment by josephj11** [comment]
## Installing The develop Branch for Testing The Latest Changes (Beyond Beta)  

Initial setup

Make an empty directory for the develop code (path/to/autokey), then 

```
git clone https://github.com/autokey/autokey path/to/autokey
git -C path/to/autokey checkout beta

```

To clone develop daily, make a cron job that runs

`git -C path/to/autokey pull
`

**Reply by BlueDrink9** [reply]
You can have a `pip install` of this that updates when you pull the latest changes by running `pip install --user -e path/to/autokey`. The `-e` flag tells pip to symlink the relevant files rather than copy them.

**Reply by ineuw** [reply]
> pip install --user -e path/to/autokey

Can you please paste an example for the path?

**Reply by BlueDrink9** [reply]
Sorry, I wasn't quite clear. This is the path to the cloned autokey source directory, e.g. (for me) `~/src/autokey`

**Comment by ineuw** [comment]
In my case, this would be path but it fails with the message below.

ineuw@LMC201DT:~$ pip install --user -e .local/lib/python3.8/site-packages/autokey

ERROR: File "setup.py" not found. Directory cannot be installed in editable mode: /home/ineuw/.local/lib/python3.8/site-packages/autokey

**Reply by BlueDrink9** [reply]
That's the installed path, not the source path. The source path would be where you clone the git repo to.

**Reply by BlueDrink9** [reply]
If you didn't clone autokey, then don't use the `-e` flag. You should just use the command to install directly from github: `pip3 install --user https://github.com/autokey/autokey/archive/v0.96.0-beta.4.zip`

**Comment by ineuw** [comment]
I didn't clone. Nevertheless, it's working fine . . . . after I cleaned up my 102 scripts. 

AutoKey has an interesting reaction to a previously found error in one of the scripts in the folder. It refuses to continue unless I fix it. even after a Restart of AK. 

These were the opening lines of `autokey-gtk --verbose`. These are X11 programs that don't exist and cannot be installed. Is this problematic? They appear in 0.96-beta4 but not in 95.10, which I am using on another hardware, same OS.

2021-06-06 16:59:24,405 INFO - autokey.UI_common_functions - Optional Commandline Program: "visgrep" was not found/able to be used correctly by AutoKey. Check that this program is correctly installed on your system.
2021-06-06 16:59:24,406 INFO - autokey.UI_common_functions - Optional Commandline Program: "png2pat" was not found/able to be used correctly by AutoKey. Check that this program is correctly installed on your system.
2021-06-06 16:59:24,406 INFO - autokey.UI_common_functions - Optional Commandline Program: "xte" was not found/able to be used correctly by AutoKey. Check that this program is correctly installed on your system.
2021-06-06 16:59:24,406 INFO - autokey.UI_common_functions - Optional Commandline Program: "xmousepos" was not found/able to be used correctly by AutoKey. Check that this program is correctly installed on your system.

**Reply by BlueDrink9** [reply]
This is a feature. It informs you if you are missing optional dependencies, whereas previously certain things would just fail when run. However, the current message is a bit too alarmist, so I will change it to something less severe.

In your case, it doesn't matter that they aren't there.

What do you mean by "don't exist and cannot be installed?"

> AutoKey has an interesting reaction to a previously found error in one of the scripts in the folder. It refuses to continue unless I fix it. even after a Restart of AK.

This sounds more serious. Could you explain this a bit more please?

**Reply by josephj11** [reply]
This last point sounds like a valid, but separate issue.

Looks like we need an addition to our wiki, maybe under "Installing", that covers the optional dependencies.

**Comment by ineuw** [comment]
Since everything worked well, after several starts, I realized that the missing apps were optional. 

------------------------

I had a collection of scripts with various errors.  I activated a script which froze Autokey and the keyboard's ability to copy and paste. I quit Autokey and restarted X, but the problem remained. X was somehow damaged and it could only be restored by a reboot. I repeated this several times to be sure. I was stymied by the error pointer, which pointed to a non-existent error in the active script. That is when I realized that Python (or Autokey) may have remembered a serious error in one of the .py files, but could not identify it correctly. 

Moved all scripts to another folder except the offending script. I retested the script several times and it worked repeatedly and flawlessly. So, I didn't report it, knowing that I must go through every script, to find the error, while I clean and standardize everything.

Spent the weekend cleaning every script using text tools of batch find and replace, then checking the results by merging all codes into a file, then run a unique sort to remove identical duplicates. The leftover revealed numerous errors which I found with the same tool and corrected it.  60% of the scripts are identical with only the variables contain different string values. I standardized the variable names to be relevant, friendly, and refer to the same object in every script. All user declared variables end with the underscore "_" to differentiate from a possible reserved word.  Now, every script functions perfectly and fast. 

The uniform contents of the scripts can be seen at https://pastebin.com/Sj9J5BYV

**Reply by josephj11** [reply]
When we enhance AutoKey so that scripts placed in the modules directory can use our API, your scripts mentioned above will be perfect candidates. You can have just one copy of each function that accepts the things that change as parameters. Then you can have new scripts that just set the parameters and call the function. This will make code maintenance much easier because you won't have to propagate changes over many copies of the same code.

**Reply by BlueDrink9** [reply]
That sounds pretty serious. Locking up clipboard input is certainly something we should never allow. I'm a little confused still by your description of how the bug was caused. Could you please give an example file and steps to reproduce it when you have time?

**Reply by josephj11** [reply]
@BlueDrink9  Yes, but as a separate issue.

**Reply by BlueDrink9** [reply]
Good point, yes. Discussions aren't great for reporting issues.

**Comment by ineuw** [comment]
@BlueDrink9, I know how exactly it happened. My first version of this loop had two errors which didn't notice until I pasted it into every script which required clearing the clipboard. 

Activating any script, crashed X windows until I repaired all and rebooted.

`i = 1`
`while i < 3:`
 `   clipboard.fill_clipboard("")`
  `  time.sleep(0.2)`
   ` i = i + 1`

I cleaned each file by opening all 100 scripts at once in a text editor and most of the errors replaced all at once, To double check,  copied all .py Autokey files into a separate folder, merged all into a single text file and sorted it uniquely. This left me with a list of every typo. It didn't take long to find and fix the errors and standardizing the variable names.

**Reply by josephj11** [reply]
It would be great if we could get back to #248. If that was implemented, then Ineuw would have had to repair just one script instead of 100! Eventually, we could develop sets of modules with common functionality that would do for AutoKey what modules do for Python now.

**Reply by josephj11** [reply]
@ineuw Why does your code need to clear the clipboard three times in a row? Doesn't it pay attention the first time? This seems like you're compensating for a bug somewhere else in the system or upstream from AutoKey.

**Comment by ineuw** [comment]
@josephj11 Found your comment by fluke. There are too many discussions in different places and there is no email notifications. 

I changed to a loop instead of using an increased time of >=0.2 seconds because the size of the data on the clipboard was not cleared and partially was appended to my copy./cut. I selected a loop because then the process is activated twice, which by experience I know it to be reliable. Since I pasted the code, I changed the code twice. from looping twice x .02 seconds, to looping three x .01 seconds. 

The delays are relevant to my work, which is limited to the max of ~7,000 characters of text which makes it easy to set a fixed delay. Looping three times in the context of my work is the best assurance so far with my level of knowledge.

**Reply by josephj11** [reply]
I understand that what you're doing works. The problem is that it should be unnecessary to clear the clipboard more than once. This means that your code is compensating for a bug somewhere else in the system or upstream.

**Comment by ineuw** [comment]
@josephj11, I agree with you that something is wrong, but very much doubt that it's Autokey. I am becoming more and more convinced that the problem is the X11 clipboard scheme. It looks to me that the PRIMARY memory used by mouse commands can contain information from an unrelated operation, while the CLIPBOARD used by Autokey is also storing information. The fact that my script pasted unrelated text makes me suspect my previous mouse use (unrelated to Autokey) may have stored information.

Mouse vs keyboard role in cut/copy/paste, and X11 clipboard scheme may cause us the problems in the context of Autokey. 

The loop is not necessarily better than a ~0.5 delay, and that is what I am experimenting with now. Shorter delay was insufficient and the script failed. Since the text length is always the same, it's easy to observe the consistency of results of different approaches.

**Comment by ineuw** [comment]
In a newly installed Linux Mint which never had Autokey, do I install 95.10 before upgrading to beta 6?

**Reply by BlueDrink9** [reply]
No, please go straight to the beta. It should work, but let us know if something doesn't install properly!

**Comment by ineuw** [comment]
The short version is that after installing it with `pip3 install --user https://github.com/autokey/autokey/archive/v0.96.0-beta.6.zip`, I damaged the installation because the "About" still displayed beta-5. 

Reinstalled it but the app is missing from  the cinnamon menu. I listed the following files of autokey-gtk in the $HOME folder. These shortcuts report an error without additional info. 

`/home/ineuw/.local/share/applications/autokey-gtk.desktop`
`/home/ineuw/.local/share/man/man1/autokey-gtk.1`
`/usr/share/app-install/desktop/autokey-gtk:autokey-gtk.desktop`

The attached file lists every file with the name "autokey-"

[autokey-.txt](https://github.com/autokey/autokey/files/6851671/autokey-.txt)

**Reply by BlueDrink9** [reply]
You didn't damage it; I made a mistake in my haste to get beta 6 out. I did it purely through github, which doesn't update the metadata in the codebase so the app still showed the old version. I'll fix it for beta 7... and I've added a test to stop myself doing that again

**Reply by ineuw** [reply]
I tried reinstalling it but I am getting the following error which may have nothing to do Autokey.

ERROR: HTTP error 404 while getting https://github.com/autokey/autokey/archive/v0.96.0-beta.6.zip.
ERROR: Could not install requirement https://github.com/autokey/autokey/archive/v0.96.0-beta.6.zip. because of error 404 Client Error: Not Found for url: https://github.com/autokey/autokey/archive/v0.96.0-beta.6.zip.
ERROR: Could not install requirement https://github.com/autokey/autokey/archive/v0.96.0-beta.6.zip. because of HTTP error 404 Client Error: Not Found for url: https://github.com/autokey/autokey/archive/v0.96.0-beta.6.zip. for URL https://github.com/autokey/autokey/archive/v0.96.0-beta.6.zip.

**Reply by BlueDrink9** [reply]
That's an issue with your local machine, because clicking that link works fine.

**Reply by ineuw** [reply]
Thanks for confirming. It is a very odd situation, but had ask to be sure. Apologies for bothering you.

This is an occasional ISP problem. I think that this is leading to a divorce after 20 years of marriage to the same ISP.

**Reply by ineuw** [reply]
Just to close the issue. It's pip3 that cannot find it. If I use `pip3 install --user https://github.com/autokey/autokey/archive/v0.96.0-beta.6.zip` then I get the error. If I paste the link in the browser, it automatically downloads the package. So, I used 
`pip3 install --user /home/ineuw/Desktop/autokey-0.96.0-beta.6` locally and it's  working fine. My thanks for your extraordinary efforts.

**Comment by ineuw** [comment]
This is a question about https://github.com/autokey/autokey/pull/586#event-5065395028, whether these corrections, adjustments, enhancements, etc., for the 95. 10, are incorporated in the current beta version as well?

**Reply by BlueDrink9** [reply]
They are not currently in the beta or 95.10. They are currently only in develop.

**Reply by ineuw** [reply]
Thanks.

**Comment by ineuw** [comment]
I love asking silly questions. I want to upgrade from -beta5 to -beta7. How do I uninstall with pip3?  `sudo pip3 uninstall autokey` doesn't work, nor does adding the usual -gtk, -qt, or the beta info? I prefer to uninstall it, but the time lost without specs is silly. (The grandchildren forbade to use the word 'stupid'.)

**Reply by ineuw** [reply]
No apologies. in pip3 it's Autokey not autokey.

**Comment by ineuw** [comment]
On a new install of Linux Mint Cinnamon. 20.2, I installed beta-7 and when I clicked on Autokey, I got the following error. Although python 3.8 is installed everywhere except in the home folder. 
![210809_0510_Autokey_96-beta7_error](https://user-images.githubusercontent.com/10367016/128693232-bcc4e046-0349-4f89-bdf1-f93fb588f44a.png)

**Reply by ineuw** [reply]
After several reboots, it started working. It seems to be a Cinnamon problem.

**Reply by josephj11** [reply]
We had this problem a few weeks ago. I thought @BlueDrink9 fixed it. #578 contains a workaround near the end. It appears to have something to do with building the Beta from the source tree.

Rebooting shouldn't really fix things like this unless there were package updates in between.

**Reply by BlueDrink9** [reply]
I have not fixed it. I am surprised that rebooting fixes it, which makes me wonder if another version of autokey ended up taking over.

**Reply by ineuw** [reply]
@josephj11 You made a good point. There were no updates. The Cinnamon Update Mananger itself has bugs, and already reported it to Linux Mint. I posted this as additional info, because rebooting is unheard of with a Linux app installation. The problem is definitely not Autokey. I used this beta 7 version in Cinnamon 20.1 without any problems. 

Reboot was necessary after other app installations as well. Often, several reboots were required to correct the app's listing in the menu. 

One problem is that the .desktop files of some apps are disabled, but not uniformly. These must share something in common which ties it to Cinnamon. I am  detailing these issues on the Linux Mint Forum.

**Comment by ineuw** [comment]
Since my last post, I reached the same productivity and versatility in Linux as in Windows. Autokey is the major part of this because I use it to generate a large variety of Wikipedia markup templates. I was using Linux Mint with Autokey 96beta.7 and last week upgraded to .9. and everything is great. My thanks to @BlueDrink9, and all contributors. 

During this period of silence, I found that Autokey and Python work perfectly well, but they are slowed down by Xorg's clipboard's concept and execution, where Autokey is concerned.

I spent a lot of time experimenting with **xsel** and  then **clipboard.fill_clipboard("")**. They performed equally, in that regardless of the length of .sleep(), failed to clear the clipboard half the time.

That is why I used a loop to trigger clearing more than once because it was an acceptable solution to achieve certainty.

I experimented with all the variables. Loops, delay with various lengths of text, and arrived at ~90% success rate of clearing the clipboard, The other 10% is corrected with a shell script  of **xsel -cd** which resets the clipboard for for use by Autokey scripts.

Also experimented with cut/paste and copy/paste blocks of formatted text. In the context of my work, cut/paste seems more reliable. Is this possible?

**Comment by josephj11** [comment]
Thanks for this painstaking work. It is probably something our developers would have been unlikely to focus on.

The cut vs. copy appears to be an anomaly or bug in upstream software. It makes no logical sense to me.

This whole thing of the clipboard not clearing the first time, every time really surprises me because if it is indeed pervasive beyond the realm of AutoKey (and I can't see why it wouldn't be), one would have expected a deluge of problem reports from all over the Linux community followed by some swift resolution.

I don't travel in X11 and similar support and development circles, but I expect I would have seen something about this somewhere besides AutoKey, but I have not.

Once you have completed your research, it would be great if you could add to or update our wiki to reflect what you have learned.

You might also consider (if you haven't already) making some sort of package/project out of the templates and tools you have developed so others can easily adopt them. Aside from being a useful resource, this might also drive more users to AutoKey and might even attract the attention of some systems level developers and motivate them to look at the underlying issues.
